lib/grandpa/grandpa.go

```diff
@@ -1281,3 +1372,208 @@ func (s *Service) lenVotes(stage Subround) int {
 
 	return count
 }
+
+func (s *Service) handleVoteMessage(from peer.ID, vote *VoteMessage) {
+	logger.Debugf("received vote message, (peer: %s, msg: %+v)", from, vote)
+	s.sendTelemetryVoteMessage(vote)
+
+	v, err := s.validateVoteMessage(from, vote)
+	if err != nil {
+		logger.Debugf("failed to validate vote message {%v}: %s", vote, err)
+		return
+	}
+
+	logger.Debugf(
+		"validated vote message %v from %s, round %d, subround %d, "+
+			"prevote count %d, precommit count %d, votes needed %d",
+		v, vote.Message.AuthorityID, vote.Round, vote.Message.Stage,
+		s.lenVotes(prevote), s.lenVotes(precommit), s.state.threshold()+1)
+}
+
+func (s *Service) handleCommitMessage(commitMessage *CommitMessage) error {
+	logger.Debugf("received commit message, msg: %+v", commitMessage)
+
+	err := verifyBlockHashAgainstBlockNumber(s.blockState,
+		commitMessage.Vote.Hash, uint(commitMessage.Vote.Number))
+	if err != nil {
+		if errors.Is(err, chaindb.ErrKeyNotFound) {
+			s.tracker.addCommit(commitMessage)
+			logger.Infof("we might not have synced to the given block %s yet: %s",
+				commitMessage.Vote.Hash, err)
+			return nil
+		}
+
+		return fmt.Errorf("verifying block hash against block number: %w", err)
+	}
+
+	containsPrecommitsSignedBy := make([]string, len(commitMessage.AuthData))
+	for i, authData := range commitMessage.AuthData {
+		containsPrecommitsSignedBy[i] = authData.AuthorityID.String()
+	}
+
+	s.telemetry.SendMessage(
+		telemetry.NewAfgReceivedCommit(
+			commitMessage.Vote.Hash,
+			fmt.Sprint(commitMessage.Vote.Number),
+			containsPrecommitsSignedBy,
+		),
+	)
+
+	has, err := s.blockState.HasFinalisedBlock(commitMessage.Round, s.state.setID)
+	if err != nil {
+		return fmt.Errorf("checking if there is finalized block: %w", err)
+	}
+
+	if has {
+		return nil
+	}
+
+	// check justification here
+	err = verifyCommitMessageJustification(commitMessage, s.state.setID,
+		s.state.threshold(), s.authorities(), s.blockState)
+	if err != nil {
+		if errors.Is(err, blocktree.ErrStartNodeNotFound) {
+			// we haven't synced the committed block yet, add this to the tracker for later processing
+			s.tracker.addCommit(commitMessage)
```

Comment:
we should `return nil` in this if block

Reply:
Not really, an error occurs in the function `verifyCommitMessageJustification` and if the error is `blocktree.ErrStartNodeNotFound` then we track the commit message to a future retry (managed by `commit_tracker`) and we should return an error to the caller the commit message could not be verified correctly otherwise we are telling to the caller the commit message is ok which is not correct.